packages/flutter/lib/src/cupertino/loupe.dart

```diff
@@ -0,0 +1,266 @@
+            CupertinoLoupe.kSize.width / 2,
+        verticalPositionOfLens -
+            (CupertinoLoupe.kSize.height -
+                CupertinoLoupe.kVerticalFocalPointOffset));
+
+    final Rect screenRect = Offset.zero & MediaQuery.of(context).size;
+
+    // Adjust the loupe position so that it never exists outside the horizontal
+    // padding.
+    final Offset adjustedLoupePosition = LoupeController.shiftWithinBounds(
+      bounds: Rect.fromLTRB(
+          screenRect.left +
+              CupertinoTextEditingLoupe._kHorizontalScreenEdgePadding,
+          // iOS doesn't reposition for Y, so we should expand the threshold
+          // so we can send the whole loupe out of bounds if need be.
+          screenRect.top -
+              (CupertinoLoupe.kSize.height +
+                  CupertinoLoupe.kVerticalFocalPointOffset),
+          screenRect.right -
+              CupertinoTextEditingLoupe._kHorizontalScreenEdgePadding,
+          screenRect.bottom +
+              (CupertinoLoupe.kSize.height +
+                  CupertinoLoupe.kVerticalFocalPointOffset)),
+      rect: rawLoupePosition & CupertinoLoupe.kSize,
+    ).topLeft;
+
+    setState(() {
+      _currentAdjustedLoupePosition = adjustedLoupePosition;
+      // the lens should always point to the center of the line.
+      _verticalFocalPointAdjustment =
+          verticalPositionOfLens - verticalCenterOfCurrentLine;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedPositioned(
+      duration: CupertinoTextEditingLoupe._kDragAnimationDuration,
+      curve: Curves.easeOut,
+      left: _currentAdjustedLoupePosition.dx,
+      top: _currentAdjustedLoupePosition.dy,
+      child: CupertinoLoupe(
+        ioAnimation: _ioAnimation,
+        additionalFocalPointOffset: Offset(0, _verticalFocalPointAdjustment),
+      ),
+    );
+  }
+}
+
+/// A [RawLoupe] in the Cupertino style.
+///
+/// Control the position and display status of [CupertinoLoupe]
+/// through a given [LoupeController].
+///
+/// [CupertinoLoupe] is a wrapper around [RawLoupe] that handles styling
+/// and transitions.
+///
+/// See also:
+/// * [RawLoupe], the backing implementation.
+/// * [CupertinoTextEditingLoupe], a widget that positions [CupertinoLoupe] based on
+/// [LoupeSelectionOverlayInfoBearer].
+/// * [LoupeController], the controller for this loupe.
+class CupertinoLoupe extends StatelessWidget {
+  /// Creates a [RawLoupe] in the Cupertino style.
+  const CupertinoLoupe({
+    super.key,
+    this.additionalFocalPointOffset = Offset.zero,
+    this.ioAnimation,
+  });
+  // These constants were eyeballed on an iPhone XR iOS v15.5.
+
+  @visibleForTesting
+  /// The vertical offset, from the center of the loupe,
+  /// that the focal point should point to.
+  static const double kVerticalFocalPointOffset = -25;
+
+  @visibleForTesting
+  /// The size of the loupe.
+  static const Size kSize = Size(82.5, 45);
```

Comment:
Todo: switch these from constants to default parameters, so users can override them if need be